src/reflect/scala/reflect/internal/tpe/GlbLubs.scala

```diff
@@ -83,69 +89,101 @@ private[internal] trait GlbLubs {
         logResult("Retracting dummies from " + tp + " in lublist")(tp.typeConstructor)
       case _ => tp
     }
-    // pretypes is a tail-recursion-preserving accumulator.
-    @tailrec
-    def loop(pretypes: List[Type], tsBts: List[List[Type]]): List[Type] = {
-      lubListDepth = lubListDepth.incr
 
-      if (tsBts.isEmpty || (tsBts exists typeListIsEmpty)) pretypes.reverse
-      else if (tsBts.tail.isEmpty) pretypes.reverse ++ tsBts.head
-      else {
+    val baseTypeSeqs: Array[BaseTypeSeq] = mapToArray(ts)(_.baseTypeSeq)
+    val ices: Array[Int] = new Array[Int](baseTypeSeqs.length)
+
+    def printLubMatrixAux(depth: Depth): Unit = {
+      val btsMap: Map[Type, List[Type]] = ts.zipWithIndex.map {
+        case (ty, ix) => ty -> baseTypeSeqs(ix).toList.drop(ices(ix))
+      }.toMap
+      printLubMatrix(btsMap, depth)
+    }
+
+    def headOf(ix: Int) = baseTypeSeqs(ix).rawElem(ices(ix))
+
+    val pretypes: mutable.ListBuffer[Type] = mutable.ListBuffer.empty[Type]
+
+    var isFinished = false
+    while (! isFinished && ices(0) < baseTypeSeqs(0).length){
+      lubListDepth = lubListDepth.incr
+      // Step 1: run through the List with these variables:
+      // 1) Is there any empty list? Are they equal or are we taking the smallest?
+      // isFinished: tsBts.exists(typeListIsEmpty)
+      // Is the frontier made up of types with the same symbol?
+      var isUniformFrontier =  true
+      var sym = headOf(0).typeSymbol
+      // var tsYs = tsBts
+      var ix = 0
+      while (! isFinished && ix < baseTypeSeqs.length){
+        if (ices(ix) == baseTypeSeqs(ix).length)
+          isFinished = true
+        else {
+          val btySym = headOf(ix).typeSymbol
+          isUniformFrontier = isUniformFrontier && sym.id == btySym.id
+          if (btySym isLess sym)
+            sym = btySym
+        }
+        ix += 1
+      }
+      // Produce a single type for this frontier by merging the prefixes and arguments of those
+      // typerefs that share the same symbol: that symbol is the current maximal symbol for which
+      // the invariant holds, i.e., the one that conveys most information regarding subtyping. Before
+      // merging, strip targs that refer to bound tparams (when we're computing the lub of type
+      // constructors.) Also filter out all types that are a subtype of some other type.
+      if (! isFinished){
         // ts0 is the 1-dimensional frontier of symbols cutting through 2-dimensional tsBts.
         // Invariant: all symbols "under" (closer to the first row) the frontier
         // are smaller (according to _.isLess) than the ones "on and beyond" the frontier
-        val ts0 = tsBts map (_.head)
-
         // Is the frontier made up of types with the same symbol?
```

Comment:
"// Is the frontier made up of types with the same symbol?" comment has been duplicated, remove it here.